firedrake/preconditioners/facet_split.py

```diff
@@ -0,0 +1,293 @@
+from firedrake.petsc import PETSc
+from firedrake.preconditioners.base import PCBase
+import firedrake.dmhooks as dmhooks
+import numpy
+
+
+__all__ = ['FacetSplitPC']
+
+
+class FacetSplitPC(PCBase):
+    """ A preconditioner that splits a function into interior and facet DOFs.
+
+    Internally this creates a PETSc PC object that can be controlled
+    by options using the extra options prefix ``facet_``.
+
+    This allows for statically-condensed preconditioners to be applied to
+    linear systems involving the matrix applied to the full set of DOFs. Code
+    generated for the matrix-free operator evaluation in the space with full
+    DOFs will run faster than the one with interior-facet decoposition, since
+    the full element has a simpler structure.
+    """
+
+    needs_python_pmat = False
+    _prefix = "facet_"
+
+    _permutation_cache = {}
+
+    def get_permutation(self, V, W):
+        from mpi4py import MPI
+        key = (V, W)
+        if key not in self._permutation_cache:
+            indices = get_permutation_map(V, W)
+            if V.comm.allreduce(numpy.all(indices[:-1] <= indices[1:]), MPI.PROD):
+                self._permutation_cache[key] = None
+            else:
+                self._permutation_cache[key] = indices
+        return self._permutation_cache[key]
+
+    def initialize(self, pc):
+        from ufl import RestrictedElement, MixedElement, TensorElement, VectorElement
+        from firedrake import FunctionSpace, TestFunctions, TrialFunctions
+        from firedrake.assemble import allocate_matrix, TwoFormAssembler
+        from functools import partial
```

Comment:
Move (all) imports to the top of the file

Reply:
I get a circular import error if I do that for `firedrake.assemble`.

Reply:
Ah, leave the assemble where it is then!